src/components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,155 @@
+import React, { useEffect, useRef, useState } from 'react';
+import classNames from 'classnames';
+import { Todo } from '../../types/Todo';
+import { updateTodo } from '../../api/todos';
+
+type Props = {
+  todo: Todo,
+  processingIds: number[],
+  onDelete: (id: number) => void,
+  onToggle: (
+    todoId: number,
+    isCompleted: boolean,
+  ) => Promise<void>;
+  togglingId: number | null,
+  onUpdate: (todoId: number, data: Todo) => void,
+  areSubmiting: boolean,
+};
+
+export const TodoItem: React.FC<Props> = ({
+  todo,
+  processingIds,
+  onDelete,
+  onToggle,
+  togglingId,
+  onUpdate,
+  areSubmiting,
+}) => {
+  const [isEditing, setIsEditing] = useState(false);
+  const [newTodoTitle, setNewTodoTitle] = useState(todo.title);
+
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    if (isEditing) {
+      inputRef.current?.focus();
+    }
+  }, [isEditing]);
+
+  const handleTodoTitleChange = () => {
+    setIsEditing(true);
+  };
+
+  const handleFormSubmit = (event: React.FormEvent) => {
+    event.preventDefault();
+
+    const currentTodoTitle = todo.title;
+
+    if (newTodoTitle === currentTodoTitle) {
+      setIsEditing(false);
+    } else if (newTodoTitle.trim() === '') {
+      onDelete(todo.id);
+    } else {
+      setIsEditing(true);
+      const todoUpdateData = { title: newTodoTitle };
+
+      updateTodo(todo.id, todoUpdateData)
+        .then((updatedTodo) => {
+          setIsEditing(false);
+          onUpdate(updatedTodo.id, updatedTodo);
```

Comment:
Why you update todo and then again call another function that also make a `PATCH` request